papermill/engines.py

```diff
@@ -0,0 +1,328 @@
+
+        self.save()
+
+    @catch_nb_assignment
+    def cell_exception(self, cell, **kwargs):
+        """
+        Called by engines when an exception is raised within a notebook to
+        set the metadata on the notebook indicating the location of the
+        failure.
+        """
+        cell.metadata.papermill['exception'] = True
+        cell.metadata.papermill['status'] = self.FAILED
+        self.nb.metadata.papermill['exception'] = True
+
+    @catch_nb_assignment
+    def cell_complete(self, cell, **kwargs):
+        """
+        Optionally called by engines during execution to finalize the
+        metadata for a cell and save the notebook to the output path.
+        """
+        end_time = self.now()
+        cell.metadata.papermill['end_time'] = end_time.isoformat()
+        if cell.metadata.papermill.get('start_time'):
+            start_time = dateutil.parser.parse(cell.metadata.papermill['start_time'])
+            cell.metadata.papermill['duration'] = (end_time - start_time).total_seconds()
+        if cell.metadata.papermill['status'] != self.FAILED:
+            cell.metadata.papermill['status'] = self.COMPLETED
+
+        self.save()
+        if self.pbar:
+            self.pbar.update(1)
+
+    @catch_nb_assignment
+    def notebook_complete(self, **kwargs):
+        """
+        Finalizes the metadata for the notebook and saves the notebook to
+        the output path.
+
+        Called by EngineBase when execution concludes, regardless of exceptions.
+        """
+        self.end_time = self.now()
+        self.nb.metadata.papermill['end_time'] = self.end_time.isoformat()
+        if self.nb.metadata.papermill.get('start_time'):
+            self.nb.metadata.papermill['duration'] = (
+                self.end_time - self.start_time
+            ).total_seconds()
+
+        # Cleanup cell statuses in case callbacks were never called
+        for cell in self.nb.cells:
+            if cell.metadata.papermill['status'] == self.FAILED:
+                break
+            elif cell.metadata.papermill['status'] == self.PENDING:
+                cell.metadata.papermill['status'] = self.COMPLETED
+
+        self.complete_pbar()
+        self.cleanup_pbar()
+
+        # Force a final sync
+        self.save()
+
+    def complete_pbar(self):
+        if hasattr(self, 'pbar') and self.pbar:
+            self.pbar.n = len(self.nb.cells)
+            self.pbar.refresh()
+
+    def cleanup_pbar(self):
+        if hasattr(self, 'pbar') and self.pbar:
+            self.pbar.close()
+            self.pbar = None
+
+    def __del__(self):
+        self.cleanup_pbar()
+
+
+class EngineBase(object):
+    """
+    Base class from which engines should inherit and implement execute_notebook.
+    Defines wrap_and_execute_notebook method which is used to correctly setup
+    the EngineNotebookWrapper object for engines to interact against.
+    """
+
+    @classmethod
+    def wrap_and_execute_notebook(
```

Comment:
Can we swap the names of `wrap_and_execute_notebook` and `execute_notebook`?

Somehow `execute_notebook` is the thing I was expecting to interact with as a user of an `Engine` instance. But now I think actually I'd be using `wrap_and_execute_notebook`.

I propose `execute_notebook` (new name of `wrap_and_execute_notebook) as the entry point for users, and `execute_wrapped_notebook` (new name for `execute_notebook`) as the thing engine implementors have to implement.

Just feels weird to be using the function with the complex name and as a user "I don't want my notebook wrapped, just executed please"

WDYT?

Reply:
That's a good point. I maybe optimized the naming for engine developers as they just needed to focus on `execute_notebook` implementation. I'm good with potentially swapping the names the way suggested here.